### ISSUE TITLE
Memory leak issue with traefik 2.2.1

### ISSUE THREAD

**Issue by sumit-joshi-mt** [traefik/traefik#7406]
### Do you want to request a *feature* or report a *bug*?
Bug

### What did you do?

- Running Traefik as Deployment (ingress controller) in K8s
- Having Bunch of services running where traefik serves as ingress controller

### What did you expect to see?
Traefik to work and reply with error when tried to access such a service

### What did you see instead?
Traefik pods memory is increasing gradually and eventually being killed by OOM killer

### Output of `traefik version`: (_What version of Traefik are you using?_)

Version:      2.2.1
Codename:     chevrotin
Go version:   go1.14.2
Built:        2020-04-29T18:02:09Z
OS/Arch:      linux/amd64

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```Name:         traefik-69949cd84f-qghtk
Namespace:    traefik
Priority:     0
Node:         ip-10-83-69-54.ap-south-1.compute.internal/10.83.69.54
Start Time:   Fri, 09 Oct 2020 14:45:24 +0530
Labels:       app.kubernetes.io/instance=traefik
              app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=traefik
              helm.sh/chart=traefik-8.4.1
              pod-template-hash=69949cd84f
Annotations:  kubernetes.io/psp: eks.privileged
              sumologic.com/include: true
Status:       Running
IP:           10.83.87.234
IPs:
  IP:           10.83.87.234
Controlled By:  ReplicaSet/traefik-69949cd84f
Containers:
  traefik:
    Container ID:  docker://affd1fdd9351967378a95873db9c5642dfc03f7fca777e93a47c4199ed43679d
    Image:         traefik:2.2.1
    Image ID:      docker-pullable://traefik@sha256:ad4442a6f88cf35266542588f13ae9984aa058a55a518a87876e48c160d19ee0
    Ports:         8082/TCP, 9000/TCP, 8000/TCP, 8443/TCP
    Host Ports:    0/TCP, 0/TCP, 0/TCP, 0/TCP
    Args:
      --entryPoints.metrics.address=:8082
      --entryPoints.traefik.address=:9000
      --entryPoints.web.address=:8000
      --entryPoints.websecure.address=:8443
      --api.dashboard=true
      --ping=true
      --providers.kubernetescrd
      --providers.kubernetesingress
      --providers.kubernetesingress.ingressclass=traefik
      --accesslog=true
      --accesslog.format=json
      --metrics.prometheus=true
      --metrics.prometheus.entryPoint=metrics
      --metrics.prometheus.addEntryPointsLabels=true
      --metrics.prometheus.addServicesLabels=true
      --entryPoints.metrics.address=:8082
      --metrics.prometheus.entryPoint=metrics
      --log.level=DEBUG
    State:          Running
      Started:      Fri, 09 Oct 2020 17:54:58 +0530
    Last State:     Terminated
      Reason:       OOMKilled
      Exit Code:    137
      Started:      Fri, 09 Oct 2020 14:45:26 +0530
      Finished:     Fri, 09 Oct 2020 17:54:58 +0530
    Ready:          True
    Restart Count:  1
    Limits:
      cpu:     500m
      memory:  1400Mi
    Requests:
      cpu:        300m
      memory:     1000Mi
    Liveness:     http-get http://:9000/ping delay=10s timeout=2s period=10s #success=1 #failure=3
    Readiness:    http-get http://:9000/ping delay=10s timeout=2s period=10s #success=1 #failure=1
    Environment:  <none>
    Mounts:
      /data from data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from traefik-token-wn5kj (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  data:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  traefik-token-wn5kj:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  traefik-token-wn5kj
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason     Age                  From                                                 Message
  ----     ------     ----                 ----                                                 -------
  Warning  Unhealthy  34m (x2 over 34m)    kubelet, ip-10-83-69-54.ap-south-1.compute.internal  Liveness probe failed: Get http://10.83.87.234:9000/ping: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
  Warning  Unhealthy  34m (x2 over 34m)    kubelet, ip-10-83-69-54.ap-south-1.compute.internal  Readiness probe failed: Get http://10.83.87.234:9000/ping: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
  Normal   Pulled     33m (x2 over 3h43m)  kubelet, ip-10-83-69-54.ap-south-1.compute.internal  Container image "traefik:2.2.1" already present on machine
  Normal   Created    33m (x2 over 3h43m)  kubelet, ip-10-83-69-54.ap-south-1.compute.internal  Created container traefik
  Normal   Started    33m (x2 over 3h43m)  kubelet, ip-10-83-69-54.ap-south-1.compute.internal  Started container traefik
  Warning  Unhealthy  33m                  kubelet, ip-10-83-69-54.ap-south-1.compute.internal  Readiness probe failed: Get http://10.83.87.234:9000/ping: read tcp 10.83.69.54:35202->10.83.87.234:9000: read: connection reset by peer
  Warning  Unhealthy  33m                  kubelet, ip-10-83-69-54.ap-south-1.compute.internal  Liveness probe failed: Get http://10.83.87.234:9000/ping: read tcp 10.83.69.54:35172->10.83.87.234:9000: read: connection reset by peer
  Normal   Killing    33m                  kubelet, ip-10-83-69-54.ap-south-1.compute.internal  Container traefik failed liveness probe, will be restarted
```

### Memory usage
Memory usage of pods, even if traffic decreases memory usage never drops it keeps increasing gradually - 
<img width="1377" alt="Screenshot 2020-10-09 at 6 38 18 PM" src="https://user-images.githubusercontent.com/42432808/95586818-d152cc00-0a5e-11eb-91a9-fb148b1d0df0.png">

Here dip in memory usage indicate restarts - 
<img width="1440" alt="Screenshot 2020-10-09 at 6 36 31 PM" src="https://user-images.githubusercontent.com/42432808/95586824-d44dbc80-0a5e-11eb-8081-0d81dcb67b70.png">
<img width="1384" alt="Screenshot 2020-10-09 at 6 41 34 PM" src="https://user-images.githubusercontent.com/42432808/95586973-14ad3a80-0a5f-11eb-84de-1494f2730000.png">

I'm not able to figure out why traefik pods memory keeps increasing, traffic and number of connections are seems to be normal. I'm i missing something? Is there a way we can take heap dump of traefik?

**Comment by jbdoumenjou** [comment]
Hi @sumit-joshi-mt ,

Thank you for your interest in Traefik.

Could you test it with the last version of Traefik (2.3.1) ?
Do you think it may be related to #7384 ?

**Comment by sumit-joshi-mt** [comment]
@jbdoumenjou 
Hi, it looks like same issue (but we are using helm3 for a long time now, we started facing this issue recently), earlier traefik pods used to work fine with 400-600Mi memory, now in our case memory keeps increasing upto around 1.4GB and at this usage its remains constant no spikes. facing same issue with new version of Traefik as well (2.3.1)
Thank you

**Comment by SantoDE** [comment]
How many services you have discovered with Traefik? For a profile, I guess[pprof]( https://blog.golang.org/pprof) guess would be good :)

**Comment by sylr** [comment]
@sumit-joshi-mt could you try one of these image which contains (almost other things) a patch to discard helm objects:

- quay.io/sylr/traefik:v2.2.11_sylr.1
- quay.io/sylr/traefik:v2.3.1_sylr.1

**Comment by sylr** [comment]
@sumit-joshi-mt could you try quay.io/sylr/traefik:v2.3.1_sylr.2 ?

**Comment by sumit-joshi-mt** [comment]
Hi @sylr, apologies for delay. I have deployed this now, will get get back to you in a day or two.
Thank you.

**Comment by sumit-joshi-mt** [comment]
Hi, @sylr I have monitored traefik for more than 2 days now, memory usage seems fine with your patch.
Thanks a lot for this.

**Comment by ldez** [comment]
Closed by  #7562
